File: backend-examples/expo_push_service.py
"""
Expo Push Notification Service - Backend Implementation
Use this to send push notifications from your backend server

Prerequisites:
- pip install httpx (or use requests)
- Store user's Expo push tokens in your database

Expo Push API: https://docs.expo.dev/push-notifications/sending-notifications/
"""

import httpx
from typing import List, Optional, Dict, Any
from dataclasses import dataclass
from enum import Enum


class NotificationPriority(str, Enum):
    DEFAULT = "default"
    NORMAL = "normal"
    HIGH = "high"


class NotificationChannel(str, Enum):
    DEFAULT = "default"
    ORDERS = "orders"
    PROMOTIONS = "promotions"
    REMINDERS = "reminders"


@dataclass
class PushNotification:
    """Push notification payload"""
    to: str  # Expo push token (ExponentPushToken[xxx])
    title: str
    body: str
    data: Optional[Dict[str, Any]] = None
    sound: str = "default"
    badge: Optional[int] = None
    channel_id: Optional[str] = None  # Android channel
    priority: NotificationPriority = NotificationPriority.HIGH
    ttl: int = 86400  # Time to live in seconds (default 1 day)
    subtitle: Optional[str] = None  # iOS only


class ExpoPushService:
    """
    Service to send push notifications via Expo's Push API
    
    Usage:
        service = ExpoPushService()
        
        # Send single notification
        result = await service.send_notification(
            token="ExponentPushToken[xxxxx]",
            title="New Order!",
            body="You have a new order #1234",
            data={"type": "order", "order_id": "1234"}
        )
        
        # Send to multiple users
        results = await service.send_bulk_notifications([
            PushNotification(to="token1", title="Hello", body="World"),
            PushNotification(to="token2", title="Hello", body="World"),
        ])
    """
    
    EXPO_PUSH_URL = "https://exp.host/--/api/v2/push/send"
    
    def __init__(self):
        self.client = httpx.AsyncClient(timeout=30.0)
    
    async def send_notification(
        self,
        token: str,
        title: str,
        body: str,
        data: Optional[Dict[str, Any]] = None,
        sound: str = "default",
        badge: Optional[int] = None,
        channel_id: str = NotificationChannel.DEFAULT,
        priority: NotificationPriority = NotificationPriority.HIGH,
        subtitle: Optional[str] = None,
    ) -> Dict[str, Any]:
        """
        Send a single push notification
        
        Args:
            token: Expo push token (ExponentPushToken[xxx])
            title: Notification title
            body: Notification body text
            data: Custom data payload (for deep linking, etc.)
            sound: Sound to play ("default" or custom sound name)
            badge: Badge count for iOS
            channel_id: Android notification channel
            priority: Notification priority
            subtitle: iOS subtitle
            
        Returns:
            Response from Expo Push API
        """
        notification = PushNotification(
            to=token,
            title=title,
            body=body,
            data=data or {},
            sound=sound,
            badge=badge,
            channel_id=channel_id,
            priority=priority,
            subtitle=subtitle,
        )
        
        results = await self.send_bulk_notifications([notification])
        return results[0] if results else {"status": "error", "message": "No response"}
    
    async def send_bulk_notifications(
        self,
        notifications: List[PushNotification]
    ) -> List[Dict[str, Any]]:
        """
        Send multiple push notifications in a single request
        
        Args:
            notifications: List of PushNotification objects
            
        Returns:
            List of responses from Expo Push API
        """
        if not notifications:
            return []
        
        # Build payload
        messages = []
        for notif in notifications:
            message = {
                "to": notif.to,
                "title": notif.title,
                "body": notif.body,
                "data": notif.data or {},
                "sound": notif.sound,
                "priority": notif.priority.value,
                "ttl": notif.ttl,
            }
            
            if notif.badge is not None:
                message["badge"] = notif.badge
            if notif.channel_id:
                message["channelId"] = notif.channel_id
            if notif.subtitle:
                message["subtitle"] = notif.subtitle
                
            messages.append(message)
        
        # Send request
        try:
            response = await self.client.post(
                self.EXPO_PUSH_URL,
                json=messages,
                headers={
                    "Accept": "application/json",
                    "Accept-Encoding": "gzip, deflate",
                    "Content-Type": "application/json",
                }
            )
            response.raise_for_status()
            
            result = response.json()
            return result.get("data", [])
            
        except httpx.HTTPError as e:
            print(f"[ExpoPush] HTTP Error: {e}")
            return [{"status": "error", "message": str(e)} for _ in notifications]
        except Exception as e:
            print(f"[ExpoPush] Error: {e}")
            return [{"status": "error", "message": str(e)} for _ in notifications]
    
    async def send_to_user(
        self,
        user_tokens: List[str],
        title: str,
        body: str,
        data: Optional[Dict[str, Any]] = None,
        channel_id: str = NotificationChannel.DEFAULT,
    ) -> List[Dict[str, Any]]:
        """
        Send notification to all devices of a single user
        
        Args:
            user_tokens: List of Expo push tokens for the user
            title: Notification title
            body: Notification body
            data: Custom data payload
            channel_id: Android channel
            
        Returns:
            List of responses
        """
        notifications = [
            PushNotification(
                to=token,
                title=title,
                body=body,
                data=data,
                channel_id=channel_id,
            )
            for token in user_tokens
        ]
        
        return await self.send_bulk_notifications(notifications)
    
    async def close(self):
        """Close the HTTP client"""
        await self.client.aclose()


# ============================================
# Example FastAPI Integration
# ============================================

"""
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import List, Optional

router = APIRouter(prefix="/notifications", tags=["notifications"])

# Initialize service (do this once at app startup)
push_service = ExpoPushService()


class RegisterTokenRequest(BaseModel):
    push_token: str
    platform: str
    device_name: Optional[str] = None
    device_model: Optional[str] = None


class SendNotificationRequest(BaseModel):
    user_id: str
    title: str
    body: str
    data: Optional[dict] = None
    channel_id: str = "default"


@router.post("/registerPushToken")
async def register_push_token(request: RegisterTokenRequest, user_id: str):
    '''
    Register a device's push token for a user
    Store this in your database associated with the user
    '''
    # Example: Store in MongoDB
    await db.push_tokens.update_one(
        {"user_id": user_id, "token": request.push_token},
        {
            "$set": {
                "platform": request.platform,
                "device_name": request.device_name,
                "device_model": request.device_model,
                "updated_at": datetime.utcnow(),
            },
            "$setOnInsert": {
                "created_at": datetime.utcnow(),
            }
        },
        upsert=True
    )
    
    return {"success": True, "message": "Token registered"}


@router.post("/unregisterPushToken")
async def unregister_push_token(push_token: str, user_id: str):
    '''
    Unregister a device's push token (e.g., on logout)
    '''
    await db.push_tokens.delete_one({
        "user_id": user_id,
        "token": push_token
    })
    
    return {"success": True, "message": "Token unregistered"}


@router.post("/send")
async def send_notification(request: SendNotificationRequest):
    '''
    Send notification to a specific user
    '''
    # Get all tokens for the user
    tokens = await db.push_tokens.find(
        {"user_id": request.user_id}
    ).to_list(100)
    
    if not tokens:
        raise HTTPException(status_code=404, detail="No push tokens found for user")
    
    # Send to all user devices
    results = await push_service.send_to_user(
        user_tokens=[t["token"] for t in tokens],
        title=request.title,
        body=request.body,
        data=request.data,
        channel_id=request.channel_id,
    )
    
    return {"success": True, "results": results}


@router.post("/broadcast")
async def broadcast_notification(title: str, body: str, data: Optional[dict] = None):
    '''
    Send notification to all registered users
    '''
    # Get all tokens
    all_tokens = await db.push_tokens.distinct("token")
    
    notifications = [
        PushNotification(to=token, title=title, body=body, data=data)
        for token in all_tokens
    ]
    
    # Expo recommends sending in batches of 100
    results = []
    for i in range(0, len(notifications), 100):
        batch = notifications[i:i+100]
        batch_results = await push_service.send_bulk_notifications(batch)
        results.extend(batch_results)
    
    return {"success": True, "sent": len(results), "results": results}
"""


# ============================================
# Utility Functions
# ============================================

def is_valid_expo_token(token: str) -> bool:
    """Check if a token is a valid Expo push token"""
    return token.startswith("ExponentPushToken[") and token.endswith("]")


def extract_failed_tokens(results: List[Dict]) -> List[str]:
    """
    Extract tokens that failed from push results
    These should be removed from your database
    """
    failed = []
    for result in results:
        if result.get("status") == "error":
            details = result.get("details", {})
            if details.get("error") in ["DeviceNotRegistered", "InvalidCredentials"]:
                # Token is no longer valid - remove from DB
                failed.append(result.get("id"))
    return failed


# ============================================
# Example Usage
# ============================================

async def example_usage():
    """Example of how to use the push service"""
    
    service = ExpoPushService()
    
    try:
        # Send a single notification
        result = await service.send_notification(
            token="ExponentPushToken[xxxxxxxxxxxxxxxxxxxxxx]",
            title="🎉 New Order!",
            body="You received a new order #12345",
            data={
                "type": "new_order",
                "order_id": "12345",
                "screen": "/(drawer)/redemptions",
            },
            channel_id=NotificationChannel.ORDERS,
            priority=NotificationPriority.HIGH,
        )
        print(f"Single notification result: {result}")
        
        # Send bulk notifications
        notifications = [
            PushNotification(
                to="ExponentPushToken[token1]",
                title="Daily Summary",
                body="You earned 150 points today!",
                data={"type": "summary"},
                channel_id=NotificationChannel.DEFAULT,
            ),
            PushNotification(
                to="ExponentPushToken[token2]",
                title="Special Offer",
                body="20% off on all orders today!",
                data={"type": "offer", "screen": "/(drawer)/whats-new/whats-new-home"},
                channel_id=NotificationChannel.PROMOTIONS,
            ),
        ]
        
        results = await service.send_bulk_notifications(notifications)
        print(f"Bulk notification results: {results}")
        
        # Check for failed tokens
        failed = extract_failed_tokens(results)
        if failed:
            print(f"Failed tokens to remove: {failed}")
            # Remove these tokens from your database
        
    finally:
        await service.close()


if __name__ == "__main__":
    import asyncio
    asyncio.run(example_usage())

File: backend-examples/expoPushService.ts
/**
 * Expo Push Notification Service - Node.js/TypeScript Implementation
 * Use this to send push notifications from your backend server
 * 
 * Prerequisites:
 * - npm install expo-server-sdk
 * 
 * Expo Push API: https://docs.expo.dev/push-notifications/sending-notifications/
 */

import { Expo, ExpoPushMessage, ExpoPushTicket, ExpoPushReceipt } from 'expo-server-sdk';

// Notification channels (must match Android channels defined in the app)
export enum NotificationChannel {
  DEFAULT = 'default',
  ORDERS = 'orders',
  PROMOTIONS = 'promotions',
  REMINDERS = 'reminders',
}

// Notification types for deep linking
export enum NotificationType {
  NEW_ORDER = 'new_order',
  ORDER_UPDATE = 'order_update',
  POINTS_EARNED = 'points_earned',
  REDEMPTION = 'redemption',
  OFFER = 'offer',
  REMINDER = 'reminder',
  GENERAL = 'general',
}

export interface NotificationData {
  type: NotificationType;
  screen?: string;
  params?: Record<string, any>;
  [key: string]: any;
}

export interface SendNotificationParams {
  token: string;
  title: string;
  body: string;
  data?: NotificationData;
  channelId?: NotificationChannel;
  sound?: 'default' | null;
  badge?: number;
  subtitle?: string; // iOS only
  priority?: 'default' | 'normal' | 'high';
  ttl?: number; // Time to live in seconds
}

export interface BulkNotificationResult {
  successful: number;
  failed: number;
  tickets: ExpoPushTicket[];
  invalidTokens: string[];
}

/**
 * Expo Push Notification Service
 */
export class ExpoPushService {
  private expo: Expo;

  constructor() {
    this.expo = new Expo();
  }

  /**
   * Validate if a token is a valid Expo push token
   */
  isValidToken(token: string): boolean {
    return Expo.isExpoPushToken(token);
  }

  /**
   * Send a single push notification
   */
  async sendNotification(params: SendNotificationParams): Promise<ExpoPushTicket> {
    const results = await this.sendBulkNotifications([params]);
    return results.tickets[0];
  }

  /**
   * Send notifications to multiple tokens
   */
  async sendBulkNotifications(
    notifications: SendNotificationParams[]
  ): Promise<BulkNotificationResult> {
    const messages: ExpoPushMessage[] = [];
    const invalidTokens: string[] = [];

    // Build messages and filter invalid tokens
    for (const notif of notifications) {
      if (!this.isValidToken(notif.token)) {
        console.warn(`[ExpoPush] Invalid token: ${notif.token}`);
        invalidTokens.push(notif.token);
        continue;
      }

      const message: ExpoPushMessage = {
        to: notif.token,
        title: notif.title,
        body: notif.body,
        data: notif.data || {},
        sound: notif.sound ?? 'default',
        priority: notif.priority ?? 'high',
        ttl: notif.ttl ?? 86400,
        channelId: notif.channelId ?? NotificationChannel.DEFAULT,
      };

      if (notif.badge !== undefined) {
        message.badge = notif.badge;
      }
      if (notif.subtitle) {
        message.subtitle = notif.subtitle;
      }

      messages.push(message);
    }

    if (messages.length === 0) {
      return {
        successful: 0,
        failed: invalidTokens.length,
        tickets: [],
        invalidTokens,
      };
    }

    // Chunk messages (Expo recommends max 100 per request)
    const chunks = this.expo.chunkPushNotifications(messages);
    const tickets: ExpoPushTicket[] = [];
    let successful = 0;
    let failed = invalidTokens.length;

    // Send each chunk
    for (const chunk of chunks) {
      try {
        const ticketChunk = await this.expo.sendPushNotificationsAsync(chunk);
        tickets.push(...ticketChunk);

        // Count successes and failures
        for (const ticket of ticketChunk) {
          if (ticket.status === 'ok') {
            successful++;
          } else {
            failed++;
            // Check for invalid tokens
            if (ticket.details?.error === 'DeviceNotRegistered') {
              // Find the token that failed
              const index = ticketChunk.indexOf(ticket);
              if (index >= 0 && chunk[index]) {
                invalidTokens.push(chunk[index].to as string);
              }
            }
          }
        }
      } catch (error) {
        console.error('[ExpoPush] Error sending chunk:', error);
        failed += chunk.length;
      }
    }

    return {
      successful,
      failed,
      tickets,
      invalidTokens,
    };
  }

  /**
   * Send notification to all devices of a user
   */
  async sendToUser(
    userTokens: string[],
    title: string,
    body: string,
    data?: NotificationData,
    channelId?: NotificationChannel
  ): Promise<BulkNotificationResult> {
    const notifications: SendNotificationParams[] = userTokens.map((token) => ({
      token,
      title,
      body,
      data,
      channelId,
    }));

    return this.sendBulkNotifications(notifications);
  }

  /**
   * Check receipts for sent notifications
   * Call this after some time to verify delivery status
   */
  async checkReceipts(ticketIds: string[]): Promise<Record<string, ExpoPushReceipt>> {
    const receiptIdChunks = this.expo.chunkPushNotificationReceiptIds(ticketIds);
    const receipts: Record<string, ExpoPushReceipt> = {};

    for (const chunk of receiptIdChunks) {
      try {
        const receiptChunk = await this.expo.getPushNotificationReceiptsAsync(chunk);
        Object.assign(receipts, receiptChunk);
      } catch (error) {
        console.error('[ExpoPush] Error fetching receipts:', error);
      }
    }

    return receipts;
  }

  /**
   * Get tokens that should be removed from database
   * Call this with receipts to find invalid/expired tokens
   */
  getTokensToRemove(receipts: Record<string, ExpoPushReceipt>): string[] {
    const tokensToRemove: string[] = [];

    for (const [receiptId, receipt] of Object.entries(receipts)) {
      if (receipt.status === 'error') {
        if (
          receipt.details?.error === 'DeviceNotRegistered' ||
          receipt.details?.error === 'InvalidCredentials'
        ) {
          // This token is no longer valid
          console.warn(`[ExpoPush] Invalid token for receipt ${receiptId}`);
          // Note: You'll need to track which token corresponds to which receipt
        }
      }
    }

    return tokensToRemove;
  }
}

// ============================================
// Pre-built notification templates
// ============================================

export const NotificationTemplates = {
  newOrder: (orderId: string, customerName: string): SendNotificationParams => ({
    token: '', // Fill in token
    title: '🎉 New Order!',
    body: `${customerName} placed order #${orderId}`,
    data: {
      type: NotificationType.NEW_ORDER,
      screen: '/(drawer)/redemptions',
      params: { orderId },
    },
    channelId: NotificationChannel.ORDERS,
  }),

  pointsEarned: (points: number, customerName: string): SendNotificationParams => ({
    token: '',
    title: '⭐ Points Awarded',
    body: `${customerName} earned ${points} points`,
    data: {
      type: NotificationType.POINTS_EARNED,
      screen: '/(drawer)/(tabs)/dashboard',
    },
    channelId: NotificationChannel.ORDERS,
  }),

  redemption: (redeemCode: string, offerTitle: string): SendNotificationParams => ({
    token: '',
    title: '🎁 New Redemption',
    body: `Customer wants to redeem: ${offerTitle}`,
    data: {
      type: NotificationType.REDEMPTION,
      screen: '/(drawer)/redemptions',
      params: { code: redeemCode },
    },
    channelId: NotificationChannel.ORDERS,
  }),

  offerReminder: (offerDate: string): SendNotificationParams => ({
    token: '',
    title: '📅 Offer Reminder',
    body: `Your offer goes live tomorrow (${offerDate})`,
    data: {
      type: NotificationType.OFFER,
      screen: '/(drawer)/whats-new/whats-new-home',
    },
    channelId: NotificationChannel.REMINDERS,
  }),

  dailySummary: (
    totalScans: number,
    totalPoints: number,
    totalRedemptions: number
  ): SendNotificationParams => ({
    token: '',
    title: '📊 Daily Summary',
    body: `Today: ${totalScans} scans, ${totalPoints} points, ${totalRedemptions} redemptions`,
    data: {
      type: NotificationType.GENERAL,
      screen: '/(drawer)/(tabs)/dashboard',
    },
    channelId: NotificationChannel.REMINDERS,
  }),

  promotion: (title: string, body: string): SendNotificationParams => ({
    token: '',
    title,
    body,
    data: {
      type: NotificationType.OFFER,
      screen: '/(drawer)/whats-new/whats-new-home',
    },
    channelId: NotificationChannel.PROMOTIONS,
  }),
};

// ============================================
// Example Express/Fastify Routes
// ============================================

/*
import express from 'express';
import { ExpoPushService, NotificationChannel, NotificationType, NotificationTemplates } from './expoPushService';

const router = express.Router();
const pushService = new ExpoPushService();

// Database interface (implement with your DB)
interface PushTokenDoc {
  userId: string;
  token: string;
  platform: 'ios' | 'android';
  deviceName?: string;
  createdAt: Date;
  updatedAt: Date;
}

// Register push token
router.post('/registerPushToken', async (req, res) => {
  const { push_token, platform, device_name, device_model } = req.body;
  const userId = req.user.id; // From auth middleware

  if (!pushService.isValidToken(push_token)) {
    return res.status(400).json({ error: 'Invalid push token' });
  }

  // Upsert token in database
  await db.pushTokens.upsert({
    where: { userId_token: { userId, token: push_token } },
    create: {
      userId,
      token: push_token,
      platform,
      deviceName: device_name,
      deviceModel: device_model,
    },
    update: {
      platform,
      deviceName: device_name,
      deviceModel: device_model,
      updatedAt: new Date(),
    },
  });

  res.json({ success: true });
});

// Unregister push token
router.post('/unregisterPushToken', async (req, res) => {
  const { push_token } = req.body;
  const userId = req.user.id;

  await db.pushTokens.delete({
    where: { userId_token: { userId, token: push_token } },
  });

  res.json({ success: true });
});

// Send notification to user
router.post('/notify/user/:userId', async (req, res) => {
  const { userId } = req.params;
  const { title, body, data, channelId } = req.body;

  // Get all tokens for user
  const tokens = await db.pushTokens.findMany({
    where: { userId },
    select: { token: true },
  });

  if (tokens.length === 0) {
    return res.status(404).json({ error: 'No tokens found for user' });
  }

  const result = await pushService.sendToUser(
    tokens.map(t => t.token),
    title,
    body,
    data,
    channelId
  );

  // Remove invalid tokens from database
  if (result.invalidTokens.length > 0) {
    await db.pushTokens.deleteMany({
      where: { token: { in: result.invalidTokens } },
    });
  }

  res.json({
    success: true,
    sent: result.successful,
    failed: result.failed,
  });
});

// Broadcast to all users
router.post('/notify/broadcast', async (req, res) => {
  const { title, body, data, channelId } = req.body;

  // Get all unique tokens
  const tokens = await db.pushTokens.findMany({
    select: { token: true },
    distinct: ['token'],
  });

  const notifications = tokens.map(t => ({
    token: t.token,
    title,
    body,
    data,
    channelId,
  }));

  const result = await pushService.sendBulkNotifications(notifications);

  // Cleanup invalid tokens
  if (result.invalidTokens.length > 0) {
    await db.pushTokens.deleteMany({
      where: { token: { in: result.invalidTokens } },
    });
  }

  res.json({
    success: true,
    total: tokens.length,
    sent: result.successful,
    failed: result.failed,
  });
});

// Send order notification using template
router.post('/notify/new-order', async (req, res) => {
  const { sellerId, orderId, customerName } = req.body;

  const tokens = await db.pushTokens.findMany({
    where: { userId: sellerId },
  });

  const template = NotificationTemplates.newOrder(orderId, customerName);
  
  const notifications = tokens.map(t => ({
    ...template,
    token: t.token,
  }));

  const result = await pushService.sendBulkNotifications(notifications);

  res.json({ success: true, ...result });
});

export default router;
*/

// ============================================
// Singleton instance for easy import
// ============================================

export const pushService = new ExpoPushService();
export default pushService;
